CMSgov/dpc-app
# Basic integration with login.gov

## 🎫 Ticket

https://jira.cms.gov/browse/DPC-3766

## 🛠 Changes

- login_dot_gov controller for handling openid connect results
- Devise configured to authenticate via openid connect/login.gov
- User attributes now include provider, uid, given_name, family_name

## ℹ️ Context for reviewers

This just implements basic integration with login.gov; sending the user out and being able to find out about the user on return.

## ✅ Acceptance Validation

To Test

1. `make portal && make start-portal`
2. http://localhost:3100/portal/users/sign_in 
<img width="456" alt="login" src="https://github.com/CMSgov/dpc-app/assets/145699825/365eb4d0-e745-4bb3-9be4-8b3ea94dffbe">

3. Click "Sign in with OpenIDConnect"
4. Confirm go to sign in for DPC Portal (local) page 
<img width="643" alt="idp login" src="https://github.com/CMSgov/dpc-app/assets/145699825/a4668abd-fb1c-4342-af86-1e633c47d85d">

5. Create an account and go through IdP 1
6. Once you reach the "Get started verifying your identity", click "Cancel" link at bottom of page
7. Click Exit Login.gov
8. Verify at auth failure page
<img width="451" alt="failure" src="https://github.com/CMSgov/dpc-app/assets/145699825/28e0088a-df16-4ced-8fec-a182cea00020">

9. Click Try again link, then Sign in button
10. Finish IdP. Note, you can download a [yml file](https://developers.login.gov/assets/yaml/proofing.yml) (go ahead and update the values, if you want) to use for the id upload.
11. Verify you reach the success page and the name matches the name in proofing.yml
<img width="559" alt="success" src="https://github.com/CMSgov/dpc-app/assets/145699825/46d26eb1-4854-48f4-8d27-e7cdbb98ad9d">



## 🔒 Security Implications

- [x] This PR adds a new software dependency or dependencies.
- [ ] This PR modifies or invalidates one or more of our security controls.
- [x] This PR stores or transmits data that was not stored or transmitted before.
- [ ] This PR requires additional review of its security implications for other reasons.

If any security implications apply, add Jason Ashbaugh (GitHub username: StewGoin) as a reviewer and do not merge this PR without his approval.
